seccomp support for ctags

hanwen opened the pull request.

masatake commented:
Oh, thank you!

hanwen commented:
I saw some CI failures. Are you looking into them, or should I fix?

masatake commented:
>  	Always print line number in json output

This is not acceptable. The fields addedd to the output should be controled by `--fields` option.
If you need the line number in your daily work, add `--fields=+n` to your `.ctags.`.

masatake commented:
Could you add LICENSE file mentioned in the comment? I have to read it...it takes much time, I guess.

k-takata commented on main/options.h:
```diff
@@ -106,6 +106,7 @@ typedef struct sOptionValues {
 	bool machinable;		/* --machinable */
 	bool withListHeader;		/* --with-list-header */
 	bool interactive; /* --interactive */
+        bool secure; /* --_secure; implies --_interactive */
```
wrong indent.

hanwen replied:
is there is a standardized (emacs?) setting that produces your desired formatting? tabs/spaces?

masatake replied:
See http://docs.ctags.io/en/latest/contributions.html?highlight=emacs#c-language

hanwen replied:
neat, thanks!

hanwen replied:
fixed indent.

k-takata commented on main/seccomp-bpf.h:
```diff
@@ -0,0 +1,89 @@
+#define _GNU_SOURCE 1
+#endif
+
+#include <config.h>
```
https://ci.appveyor.com/project/masatake/ctags/build/1.0.2153/job/3a9rn0kq385ct8ow#L78
"config.h" is not always available. Should be surrounded by `#ifdef HAVE_CONFIG_H`...`#endif`.

hanwen replied:
done.

k-takata commented on main/seccomp.c:
```diff
@@ -0,0 +1,60 @@
+#include <stdio.h>
+#include <stddef.h>
+#include <string.h>
+#include <sys/prctl.h>
```
Compile error on Cygwin and MinGW.
https://ci.appveyor.com/project/masatake/ctags/build/1.0.2153/job/d53g4g6ylncvha6g#L268
sys/prctl.h is not available on them.

hanwen replied:
I redid this based on libseccomp, which also gets rid of the LICENSE file requirement.

k-takata commented on main/seccomp.c:
```diff
@@ -0,0 +1,60 @@
+#include <stddef.h>
+#include <string.h>
+#include <sys/prctl.h>
+#include <unistd.h>
```
VC doesn't have unistd.h.

k-takata commented on main/seccomp.c:
```diff
@@ -0,0 +1,60 @@
+#include <unistd.h>
+
+#include <stdlib.h>
+#include <sys/ucontext.h>
```
VC doesn't have sys/ucontext.h.

masatake commented on main/main.c:
```diff
@@ -496,6 +497,13 @@ static void batchMakeTags (cookedArgs *args, void *user CTAGS_ATTR_UNUSED)
 void interactiveLoop (cookedArgs *args CTAGS_ATTR_UNUSED, void *user CTAGS_ATTR_UNUSED)
 {
+        if (Option.secure) {
+                initializeParser (LANG_AUTO);
```
You are an excellent hacker!

hanwen replied:
thanks :)

masatake commented:
Could you add "seccomp" to the feature list defined in `Features` in main/options.c?
So we can write a test case for teting seccomp. I will write a test case after merging this.

hanwen commented on docs/secure-mode.rst:
oh, cut & paste error.

k-takata commented on main/options.c:
```diff
@@ -38,6 +38,7 @@
 #include "error.h"
 #include "interactive.h"
 #include "writer.h"
+#include "seccomp.h"
```
Maybe mistake of "secure.h"?

k-takata commented:
The `--help` message should be also updated.
https://github.com/universal-ctags/ctags/blob/master/main/options.c#L194-L450
This list is in alphabetical order.

hanwen commented on main/seccomp.c:
```diff
@@ -0,0 +1,51 @@
+#else
+int install_syscall_filter(void)
+{
+        fprintf(stderr, "not supported; compile with --enable-seccomp.");
```
A little bit of defensive programming: the call in interactiveLoop isn't guarded by ifdef.

(maybe this is overly generic, but there are other sandboxing mechanisms on OSX, see https://dev.chromium.org/developers/design-documents/sandbox/osx-sandboxing-design)

hanwen replied:
Done.

hanwen commented on main/options.c:
```diff
@@ -502,6 +503,9 @@ static const char *const Features [] = {
 	"interactive",
 #endif
+#ifdef HAVE_SECCOMP
+        "secure",
```
this is what c-guess came up with. Maybe it's something wrong with the binding of TAB ?

(BTW, I can recommend https://clang.llvm.org/docs/ClangFormat.html as a solution to formatting dilemmas.)

k-takata replied:
I don't know about c-guess, but it seems that it failed to guess correctly.
If you use EditorConfig, it will set the tab settings correctly.

We already have settings for Uncrustify and GNU indent. I don't know we should support clang-format besides them (or switch to it).

hanwen commented:
added the help text. Let's fix indent when everything else is perfect.

hanwen commented:
added note on osx.

k-takata commented on main/options.c:
```diff
@@ -439,6 +440,10 @@ static optionDescription LongOptionDescription [] = {
+#ifdef HAVE_SECCOMP
+ {0,"  --_secure"},
+ {0,"       Enter secure interactive mode (json over stdio, input over stdin)."},
+#endif
```
This list is in alphabetical order.
So this block should be moved between `--_list-roles` and `--_xformat`.

hanwen replied:
done, and moved --_interactive too.

hanwen commented:
added seccomp to feature list too.

masatake commented:
I tried `--_secure` option. However, ctags doesn't work; ctags stopped whep `open("/dev/urandm"' is called.

```
read(0, "{\"command\":\"generate-tags\", \"fil"..., 4096) = 79
open("/dev/urandom", O_RDONLY)          = ?
+++ killed by SIGSYS +++
```

Gdb says:
```
(gdb) r --options=NONE --_secure < INPUT
Starting program: /home/yamato/var/ctags-github/ctags --options=NONE --_secure < INPUT
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
ctags: Notice: No options will be read from files or environment
Detaching after fork from child process 8225.
{"_type": "program", "name": "Universal Ctags", "version": "0.0.0"}

Program terminated with signal SIGSYS, Bad system call.
The program no longer exists.
Missing separate debuginfos, use: dnf debuginfo-install jansson-2.9-1.fc25.x86_64 libseccomp-2.3.2-1.fc25.x86_64 libyaml-0.1.6-8.fc24.x86_64
(gdb) where
No stack.
(gdb) b main
Breakpoint 1 at 0x403260: file main/main.c, line 639.
(gdb) r
Starting program: /home/yamato/var/ctags-github/ctags --options=NONE --_secure < INPUT
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, main (argc=3, argv=0x7fffffffde18) at main/main.c:639
639	{
(gdb) b open
Breakpoint 2 at 0x7ffff64e4cb0: open. (2 locations)
(gdb) cont
Continuing.
ctags: Notice: No options will be read from files or environment
Detaching after fork from child process 8227.
{"_type": "program", "name": "Universal Ctags", "version": "0.0.0"}

Breakpoint 2, open64 () at ../sysdeps/unix/syscall-template.S:84
84	T_PSEUDO (SYSCALL_SYMBOL, SYSCALL_NAME, SYSCALL_NARGS)
(gdb) where
#0  open64 () at ../sysdeps/unix/syscall-template.S:84
#1  0x00007ffff7865a8e in json_object_seed () from /lib64/libjansson.so.4
#2  0x00007ffff7869b47 in json_object () from /lib64/libjansson.so.4
#3  0x00007ffff7866ad5 in parse_value () from /lib64/libjansson.so.4
#4  0x00007ffff7866f16 in parse_json () from /lib64/libjansson.so.4
#5  0x00007ffff786709b in json_loads () from /lib64/libjansson.so.4
#6  0x000000000040cb57 in interactiveLoop (args=<optimized out>, user=<optimized out>)
    at main/main.c:520
#7  0x00000000004032ff in runMainLoop (args=0x6a5620) at main/main.c:442
#8  main (argc=<optimized out>, argv=<optimized out>) at main/main.c:667
(gdb) step

Program terminated with signal SIGSYS, Bad system call.
The program no longer exists.

```

It seems that json_object_seed calls `open`.

I will make a pul request t you.

masatake commented on main/main.c:
```diff
@@ -496,6 +497,17 @@ static void batchMakeTags (cookedArgs *args, void *user CTAGS_ATTR_UNUSED)
+	if (Option.secure) {
+		initializeParser (LANG_AUTO);
+                // Jansson reads /dev/urandom
+                json_object_seed (0);
```
Wrong indentation. I'm sorry if it comes from my PR.

masatake commented:
Sorry for taking time for merging this.

I have two more questions.

1. As I wrote, you added line " initializeParser (LANG_AUTO); " before installing the filter.
Why? I guessed that you put the line here because xcmd based parsers call clone and execve.
If my guessing is correct, you can remove "initializeParser (LANG_AUTO); "  from the code.

2. You introduced `_secure` option. How about enabling the filter by default when `_interactive` optoin is given, and remvoing stand-alone `_secure` option?

hanwen commented:
I had to back out the json_seed change: the function is "New in version 2.6.". My ubuntu 14.04 has libjansson 2.5. 

We'll have to do an autoconf check and conditional code (sigh). Can we check the PR as proposed, and do the autoconf check in a separate change? I don't have ways to test it here, unfortunately.

regarding your questions: 

1. I think it may actually have been caused by the brk syscall which was missing in an earlier version. I removed it and it still works. (It would be nice to test with all parsers, though). 

2. in the Json mode, you can specify a filename, making ctags read the file from disk. That's obviously a no-no for a sandboxed implementation. If you want to drop that feature, we could make it the default.

masatake commented:
`secure.h` is too short. Please put it to interactive.h. `install_syscall_filter` should be installSyscallFilter.
I don't like camelcase but the convention is used in Exuberant ctags.

hanwen commented:
* I added your json seed hack
* camelCase
* add to interactive.h

i'll have a look at your suggestion for --_interactive.

hanwen commented:
I don't see an example for setMainloop with non-NULL user argument. (I would just delete that functionality). 

I am OK with your suggested behavior, but think it would be quicker if you adapted the code yourself.

masatake commented:
> I am OK with your suggested behavior, but think it would be quicker if you adapted the code yourself.

I see. I will make one more pull request.

masatake commented on main/seccomp.c:
```diff
@@ -0,0 +1,53 @@
+#ifdef HAVE_CONFIG_H
```
Could you add a copyright notice?

masatake replied:
Could you add 

```
/*
 *
 *  Copyright (c) 2015, Red Hat, Inc.
 *  Copyright (c) 2017, YOUR NAME HERE
 *
 *  Author: YOUR NAME HERE
 *
 *   This source code is released for free distribution under the terms of the
 *   GNU General Public License version 2 or (at your option) any later version.
 *
 */
```
?

hanwen replied:
done.

masatake replied:
Thank you.

masatake commented on main/main.c:
```diff
@@ -496,6 +497,21 @@ static void batchMakeTags (cookedArgs *args, void *user CTAGS_ATTR_UNUSED)
+		json_decref (tmp);
+
+		if (installSyscallFilter ()) {
+			error (FATAL, "install_syscall_filter failed");
```
exit(n) may be needed here. In interactive mode, I'm afraid 'error(FATAL' doesn't make the ctags process exit.

masatake replied:
exit(1) may be needed here. In interactive mode, I'm afraid 'error(FATAL' doesn't make the ctags process exit.  I will fix this. I have to think about the behavior of  jsonErrorPrinter in early stage.

masatake commented on main/seccomp.c:
```diff
@@ -0,0 +1,53 @@
+	int err = seccomp_load (ctx);
+	if (err < 0)
+	{
+		fprintf (stderr, "seccomp_load: %d", err);
```
`	error (WARNING,` should be used here. I will do this.
` verbose` output is also wanted. So we can see the installation of filter with --verbose option.

masatake commented:
The change I requested about option name (--_secure to --_interactive=secure) is submitted to https://github.com/hanwen/ctags/pull/4. People can give a comment the name of option. I think `--_interactive=secure` is better because it shows in which situation and usecase seccomp is enabled.
It doesn't add one more top level option. 

As soon as the copyright notice for seccomp.c is filled, I will make a branch at universal-ctags/ctags repository and merge this PR.

I will add some minor changes to the branch I commented here including adding a test case.

I think the change in this PR is one of the highlight of the planed release. I would like to see a real application of `_interactive=secure`.

hanwen commented:
I added the (c) header, fixed up some minor things, and added some TODOs.

Happy to help with polishing the docs. (I suggest "insecure" rather than "nosecure")

For an application, try

  http://cs.bazel.build/search?q=r%3Actags+option&num=50

(this currently uses ctags + namespace sandbox, but seccomp is much more secure.)

masatake commented:
Merged the original changes via #1458. 
If I missed something, please, open another case(or pull request).

@hanwen, thank you very much.

I will rename `--_interactive` to `--interactive` till releasing 6.0 or 5.9.
